subprojects/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java

```diff
@@ -202,4 +218,25 @@ public CopySpec getMetaInf() {
     public CopySpec metaInf(Closure<?> configureClosure) {
         return ConfigureUtil.configure(configureClosure, getMetaInf());
     }
+
+    /**
+     * Provides a comparator that sorts the contents of a Jar/Ear/War so that META-INF is always first.
+     */
+    private static class JarContentsComparator implements Comparator<RelativePath> {
+        @Override
+        public int compare(RelativePath o1, RelativePath o2) {
+            boolean o1MetaInf = isMetaInf(o1.getSegments()[0]);
```

Comment:
We are asserting that the `MANIFEST.MF` file is the first in the jar [here](https://github.com/gradle/gradle/blob/2a4c7794ebd5e91d102eca2f952a3d76d7d81e8b/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy#L58-L58).[The jar tool automatically generates a manifest file entry named META-INF/MANIFEST.MF. It is always the first entry in the jar file.](http://docs.oracle.com/javase/7/docs/technotes/tools/solaris/jar.html). [Some](https://grep.codeconsult.ch/2011/11/15/manifest-mf-must-be-the-first-resource-in-a-jar-file-heres-how-to-fix-broken-jars/) OSGi tools depend on that.